pkg/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,373 @@
+package cachemanager
+
+import (
+	"context"
+	"testing"
+
+	configv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/cachemanager/aggregator"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/fakes"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/watch"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/wildcard"
+	testclient "github.com/open-policy-agent/gatekeeper/v3/test/clients"
+	"github.com/open-policy-agent/gatekeeper/v3/test/testutils"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 2)
+}
+
+func unitCacheManagerForTest(t *testing.T) (*CacheManager, context.Context) {
+	cm, _, ctx := makeCacheManager(t)
+	return cm, ctx
+}
+
+func makeCacheManager(t *testing.T) (*CacheManager, client.Client, context.Context) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	cfClient := &FakeCfClient{}
+	tracker, err := readiness.SetupTracker(mgr, false, false, false)
+	require.NoError(t, err)
+	processExcluder := process.Get()
+	processExcluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []wildcard.Wildcard{"kube-system"},
+			Processes:          []string{"sync"},
+		},
+	})
+	events := make(chan event.GenericEvent, 1024)
+	w, err := wm.NewRegistrar(
+		"test-cache-manager",
+		events)
+	require.NoError(t, err)
+
+	cacheManager, err := NewCacheManager(&Config{
+		CfClient:         cfClient,
+		SyncMetricsCache: syncutil.NewMetricsCache(),
+		Tracker:          tracker,
+		ProcessExcluder:  processExcluder,
+		WatchedSet:       watch.NewSet(),
+		Registrar:        w,
+		Reader:           c,
+		GVKAggregator:    aggregator.NewGVKAggregator(),
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		ctx.Done()
+	})
+
+	t.Cleanup(func() {
+		cancelFunc()
+	})
+
+	testutils.StartManager(ctx, t, mgr)
+
+	return cacheManager, c, ctx
+}
+
+// TestCacheManager_wipeCacheIfNeeded.
+func TestCacheManager_wipeCacheIfNeeded(t *testing.T) {
```

Comment:
Organizing the `wipeCacheIfNeeded` test this way makes it hard to tell our level of coverage (must read all the tests to understand what is being tested) and means we have a lot of repeated code to test the same thing (both `wipeCacheIfNeeded` tests need to set things up similarly and test for similar effects).

It would be cleaner to examine `wipeCacheIfNeeded` using a more traditional table-driven test.  Example:

```
tcs := []struct {
  name string
  beforeCM CacheManager
  beforeCache FakeOpa  // You'll have to see if a cmp.Diff of the CacheManager also includes the fake cache.  If not, you'll have to assert its equality separately.
  wantCM CacheManager
  wantCache FakeOpa
} {
  {
    // Show that having something in gvksToDeleteFromCache yields a cache wipe
  },
  {
    // Show that setting `c.excluderChanged` yields a cache wipe
  },
  {
    // Show that when neither of the following are true, the CacheManager and cache are unchanged
  }
}

t.Run(...
  // In this section, you can set up the CacheManager to have only the data required to make assertions on if wipeCacheIfNeeded works correctly. 
```

This way, you test all the permutations of the method and can be confident that it works exactly as you intend.  Further, someone who wants to extend the code later can easily define a new test case.